### PR TITLE
ipsec: T2816: IPSec python rework, includes DMVPN and VTI support.

### PR DESCRIPTION
**Pull request by sarthurdev**
## Change Summary
Rewrite of the vyatta-cfg-vpn and vyatta-op-vpn packages into XML/Python format for vyos-1x.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [x] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T2816

## Component(s) name

ipsec

## Proposed changes

Reworks `set vpn ipsec` / `set vpn rsa-keys` and equivalent op-mode commands to XML/Python.

vyatta-cfg-vpn and vyatta-op-vpn needs to be removed as deps from vyos-world. Adds dependency on python3-crypto debian package.

## How to test

Tested against a 1.3-rc4 peer with site-to-site PSK, RSA, X509, and DMVPN, all worked correctly with identical syntax.

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: smoketest/scripts/cli/test_vpn_ipsec.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        self.cli_set(base_path + ["site-to-site", "peer", "203.0.113.45", "authentication", "mode", "pre-shared-secret"])
+        self.cli_set(base_path + ["site-to-site", "peer", "203.0.113.45", "authentication", "pre-shared-secret", "MYSECRETKEY"])
+        self.cli_set(base_path + ["site-to-site", "peer", "203.0.113.45", "ike-group", "MyIKEGroup"])
+        self.cli_set(base_path + ["site-to-site", "peer", "203.0.113.45", "default-esp-group", "MyESPGroup"])
+        self.cli_set(base_path + ["site-to-site", "peer", "203.0.113.45", "local-address", "192.0.2.10"])
+        self.cli_set(base_path + ["site-to-site", "peer", "203.0.113.45", "tunnel", "1", "protocol", "gre"])
+
+        self.cli_commit()
+
+        # TODO: verify output to config files
```
Comment (c-po) [inline smoketest/scripts/cli/test_vpn_ipsec.py]: Can you please add some output verification steps which checks that the intended config line is actually rendered?Something like: https://github.com/vyos/vyos-1x/blob/current/smoketest/scripts/cli/test_system_conntrack.py#L232-L235

**Review by c-po — CHANGES_REQUESTED** [review]
I'm not an IPSec expert but at first glance this looks really good to me. Except that small part which missed Smoketest validation.

File: src/conf_mode/vpn_rsa-keys.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import base64
+import os
+import struct
+
+from copy import deepcopy
+from subprocess import DEVNULL
```
Comment (sever-sever) [inline src/conf_mode/vpn_rsa-keys.py]: Check/delete extra exports, if you don't use it 
```
deepcopy
DEVNULL
render
pprint
```

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: src/conf_mode/vpn_rsa-keys.py (hunk 2 above)
Comment (sarthurdev) [reply src/conf_mode/vpn_rsa-keys.py]: Thanks, cleaned up.

**Comment by sarthurdev** [comment]
I have updated the smoketest to check the output.

There is a problem after DMVPN test finished and the NHRP tunnel is removed, but that seems to be a problem with the nhrp module: `Chain 'VYOS_NHRP_tun100_OUT_HOOK' does not exist`

**Review by c-po — APPROVED** [review]

**Comment by sever-sever** [comment]
I can't reset the connection.
```
vyos@r4-1.3:~$ reset vpn ipsec-peer 100.64.0.2 tunnel 0
sudo: /usr/libexec/vyos/op_mode/vpn_ipsec.py: command not found
```
Ls
```
vyos@r4-1.3:~$ sudo ls -l ${vyos_op_scripts_dir}/vpn_ipsec.py
-rw-r--r-- 1 root root 6548 Jan 11 18:02 /usr/libexec/vyos/op_mode/vpn_ipsec.py

```

**Comment by sarthurdev** [comment]
> 
> 
> I can't reset the connection.
> 
> ```
> vyos@r4-1.3:~$ reset vpn ipsec-peer 100.64.0.2 tunnel 0
> sudo: /usr/libexec/vyos/op_mode/vpn_ipsec.py: command not found
> ```
> 
> Ls
> 
> ```
> vyos@r4-1.3:~$ sudo ls -l ${vyos_op_scripts_dir}/vpn_ipsec.py
> -rw-r--r-- 1 root root 6548 Jan 11 18:02 /usr/libexec/vyos/op_mode/vpn_ipsec.py
> ```

Updated fork to add the execute bit.

File: data/templates/ipsec/ike-esp.tmpl [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    ike = {{ ike_ciphers }}
+{%      if "lifetime" in ike %}
+    ikelifetime = {{ ike.lifetime }}s
+{%      endif %}
+    reauth = {{ ike.ikev2_reauth if "ikev2_reauth" in ike else "no" }}
+    closeaction = {{ ike.close_action if "close_action" in ike else "none" }}
+{%      if "dead_peer_detection" in ike %}
+    dpdaction = {{ ike.dead_peer_detection.action }}
+    dpdtimeout = {{ ike.dead_peer_detection.timeout }}
+    dpdinterval = {{ ike.dead_peer_detection.interval }}
```
Comment (sever-sever) [inline data/templates/ipsec/ike-esp.tmpl]: Is this parameter correct?
```
 set vpn ipsec ike-group grp-IKE dead-peer-detection interval 32

vyos@r4-1.3:~$ sudo ipsec restart
Stopping strongSwan IPsec...
Starting strongSwan 5.7.2 IPsec [starter]...
# unknown keyword 'dpdinterval'
# unknown keyword 'dpdinterval'
# unknown keyword 'dpdinterval'
### 3 parsing errors (0 fatal) ###
vyos@r4-1.3:~$ 
vyos@r4-1.3:~$ sudo cat /etc/ipsec.conf | grep dpdi
    dpdinterval = 32

```
Most likely it is "dpddelay"
https://wiki.strongswan.org/projects/strongswan/wiki/ConnSection

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Comment by sever-sever** [comment]
Impossiblele to reset vpn peer 
```
vyos@r4-1.3:~$ reset vpn ipsec-peer 100.64.0.2 
no config named 'site1'
Peer reset result: success
vyos@r4-1.3:~$ 
vyos@r4-1.3:~$ show conf com | match tunnel
set vpn ipsec esp-group grp-ESP mode 'tunnel'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 allow-nat-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 allow-public-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 local prefix '10.11.10.0/24'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 remote prefix '10.21.20.0/24'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 allow-nat-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 allow-public-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 local prefix '10.12.10.0/24'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 remote prefix '10.22.20.0/24'
vyos@r4-1.3:~$ 
vyos@r4-1.3:~$ 
vyos@r4-1.3:~$ cat /etc/ipsec.conf | grep con
config setup
conn site1 # peer:100.64.0.2
conn site1t0
conn site1t1
vyos@r4-1.3:~$ 
vyos@r4-1.3:~$ sudo swanctl -l
site1t0: #1, ESTABLISHED, IKEv2, 6c7e42268a81af01_i* 88bda72e46f0f619_r
  local  '100.64.0.1' @ 100.64.0.1[500]
  remote '100.64.0.2' @ 100.64.0.2[500]
  AES_GCM_16-256/PRF_HMAC_SHA2_256/ECP_256
  established 1074s ago, rekeying in 84718s
  site1t0: #1, reqid 1, INSTALLED, TUNNEL, ESP:AES_GCM_16-256
    installed 1074s ago, rekeying in 26745s, expires in 27726s
    in  c32c8cd4,      0 bytes,     0 packets
    out c709542c,      0 bytes,     0 packets
    local  10.11.10.0/24
    remote 10.21.20.0/24
  site1t1: #2, reqid 2, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/ECP_256
    installed 1074s ago, rekeying in 26941s, expires in 27726s
    in  c9a65c4d,      0 bytes,     0 packets
    out c6d9c881,      0 bytes,     0 packets
    local  10.12.10.0/24
    remote 10.22.20.0/24
vyos@r4-1.3:~$
```

**Comment by sever-sever** [comment]
Impossible to show IKE sa
```
vyos@r4-1.3:~$ show vpn ike sa
Not implemented
vyos@r4-1.3:~$ 
```
Expected
```
vyos@r6-roll:~$ show vpn ike sa
Peer ID / IP                            Local ID / IP               
------------                            -------------
100.64.0.1                              100.64.0.2                             

    State  IKEVer  Encrypt  Hash    D-H Group      NAT-T  A-Time  L-Time
    -----  ------  -------  ----    ---------      -----  ------  ------
    up     IKEv2   aes16_256 n/a     19(ECP_256)    no     3600    86400  

 
vyos@r6-roll:~$ 

```

File: data/templates/ipsec/ipsec.conf.tmpl [hunk 4]
```diff
@@ -1,3 +1,121 @@
...
+config setup
+    charondebug = "{{ charondebug }}"
+    uniqueids = {{ "no" if disable_uniqreqids is defined else "yes" }}
+
+{%  if site_to_site is defined and "peer" in site_to_site -%}
+{%      for peer, peer_conf in site_to_site.peer.items() %}
+{%          set peer_index = loop.index %}
+{%          set peer_ike = ike_group[peer_conf.ike_group] %}
+{%          set peer_esp = esp_group[peer_conf.default_esp_group] if 'default_esp_group' in peer_conf else None %}
+conn site{{peer_index}} # peer:{{ peer }}
```
Comment (sever-sever) [inline data/templates/ipsec/ipsec.conf.tmpl]: Is it possible to use the old format for the connection name?
```
vyos@r4-1.3:~$ show vpn ipsec sa
Connection    State    Uptime    Bytes In/Out    Packets In/Out    Remote address    Remote ID    Proposal
------------  -------  --------  --------------  ----------------  ----------------  -----------  ----------------------
site1t0       up       4m56s     0B/0B           0/0               100.64.0.2        N/A          AES_GCM_16_256
site1t1       up       4m56s     0B/0B           0/0               100.64.0.2        N/A          AES_GCM_16_256/ECP_256

```
Old format:
```
vyos@r6-roll:~$ show vpn ipsec sa
Connection                State    Uptime    Bytes In/Out    Packets In/Out    Remote address    Remote ID    Proposal
------------------------  -------  --------  --------------  ----------------  ----------------  -----------  ----------------------
peer-100.64.0.1-tunnel-0  up       3m57s     0B/0B           0/0               100.64.0.1        N/A          AES_GCM_16_256
peer-100.64.0.1-tunnel-1  up       3m57s     0B/0B           0/0               100.64.0.1        N/A          AES_GCM_16_256/ECP_256
vyos@r6-roll:~$ 
```

Some customers use their own scripts for checking ipsec status match old format style.

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: data/templates/ipsec/ipsec.conf.tmpl (hunk 4 above)
Comment (sarthurdev) [reply data/templates/ipsec/ipsec.conf.tmpl]: Yes I will revert the naming now, that will also make it easier to fix the `reset vpn ipsec-peer xxxxx` script too.

**Comment by sarthurdev** [comment]
> 
> 
> Impossible to show IKE sa
> 
> ```
> vyos@r4-1.3:~$ show vpn ike sa
> Not implemented
> vyos@r4-1.3:~$ 
> ```
> 
> Expected
> 
> ```
> vyos@r6-roll:~$ show vpn ike sa
> Peer ID / IP                            Local ID / IP               
> ------------                            -------------
> 100.64.0.1                              100.64.0.2                             
> 
>     State  IKEVer  Encrypt  Hash    D-H Group      NAT-T  A-Time  L-Time
>     -----  ------  -------  ----    ---------      -----  ------  ------
>     up     IKEv2   aes16_256 n/a     19(ECP_256)    no     3600    86400  
> 
>  
> vyos@r6-roll:~$ 
> ```

Oh yes, I was hoping to commit that in a later PR as I've still to understand what is going in the vyatta version: https://github.com/vyos/vyatta-op-vpn/blob/current/lib/OPMode.pm#L1272

File: data/templates/ipsec/ipsec.conf.tmpl (hunk 4 above)
Comment (sarthurdev) [reply data/templates/ipsec/ipsec.conf.tmpl]: Reverted naming

File: data/templates/ipsec/ike-esp.tmpl (hunk 3 above)
Comment (sarthurdev) [reply data/templates/ipsec/ike-esp.tmpl]: Fixed the typo.

File: smoketest/scripts/cli/test_vpn_ipsec.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2020 VyOS maintainers and contributors
```
Comment (sever-sever) [inline smoketest/scripts/cli/test_vpn_ipsec.py]: Just a small typo  2021 in some scripts.

**Comment by sever-sever** [comment]
Such test configuration not working.

```
set vpn ipsec esp-group ESP-GRP-VTI compression 'disable'
set vpn ipsec esp-group ESP-GRP-VTI lifetime '1800'
set vpn ipsec esp-group ESP-GRP-VTI mode 'tunnel'
set vpn ipsec esp-group ESP-GRP-VTI pfs 'enable'
set vpn ipsec esp-group ESP-GRP-VTI proposal 1 encryption 'aes256'
set vpn ipsec esp-group ESP-GRP-VTI proposal 1 hash 'sha1'
set vpn ipsec esp-group grp-ESP compression 'disable'
set vpn ipsec esp-group grp-ESP lifetime '28800'
set vpn ipsec esp-group grp-ESP mode 'tunnel'
set vpn ipsec esp-group grp-ESP pfs 'dh-group19'
set vpn ipsec esp-group grp-ESP proposal 10 encryption 'aes256gcm128'
set vpn ipsec esp-group grp-ESP proposal 10 hash 'sha256'
set vpn ipsec ike-group IKE-GRP-VTI ikev2-reauth 'no'
set vpn ipsec ike-group IKE-GRP-VTI key-exchange 'ikev1'
set vpn ipsec ike-group IKE-GRP-VTI lifetime '3600'
set vpn ipsec ike-group IKE-GRP-VTI proposal 1 encryption 'aes256'
set vpn ipsec ike-group IKE-GRP-VTI proposal 1 hash 'sha1'
set vpn ipsec ike-group grp-IKE dead-peer-detection action 'hold'
set vpn ipsec ike-group grp-IKE dead-peer-detection interval '32'
set vpn ipsec ike-group grp-IKE dead-peer-detection timeout '120'
set vpn ipsec ike-group grp-IKE ikev2-reauth 'no'
set vpn ipsec ike-group grp-IKE key-exchange 'ikev2'
set vpn ipsec ike-group grp-IKE lifetime '86400'
set vpn ipsec ike-group grp-IKE mobike 'disable'
set vpn ipsec ike-group grp-IKE proposal 10 dh-group '19'
set vpn ipsec ike-group grp-IKE proposal 10 encryption 'aes256gcm128'
set vpn ipsec ike-group grp-IKE proposal 10 hash 'sha256'
set vpn ipsec ipsec-interfaces interface 'eth1'
set vpn ipsec site-to-site peer 100.64.0.2 authentication id '100.64.0.1'
set vpn ipsec site-to-site peer 100.64.0.2 authentication mode 'pre-shared-secret'
set vpn ipsec site-to-site peer 100.64.0.2 authentication pre-shared-secret 'SSSeeccRetT'
set vpn ipsec site-to-site peer 100.64.0.2 authentication remote-id '100.64.0.2'
set vpn ipsec site-to-site peer 100.64.0.2 connection-type 'initiate'
set vpn ipsec site-to-site peer 100.64.0.2 ike-group 'grp-IKE'
set vpn ipsec site-to-site peer 100.64.0.2 ikev2-reauth 'inherit'
set vpn ipsec site-to-site peer 100.64.0.2 local-address '100.64.0.1'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 allow-nat-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 allow-public-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 local prefix '10.11.10.0/24'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 0 remote prefix '10.21.20.0/24'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 allow-nat-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 allow-public-networks 'disable'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 local prefix '10.12.10.0/24'
set vpn ipsec site-to-site peer 100.64.0.2 tunnel 1 remote prefix '10.22.20.0/24'
set vpn ipsec site-to-site peer 192.0.2.2 authentication mode 'pre-shared-secret'
set vpn ipsec site-to-site peer 192.0.2.2 authentication pre-shared-secret 'SeCrEt'
set vpn ipsec site-to-site peer 192.0.2.2 ike-group 'IKE-GRP-VTI'
set vpn ipsec site-to-site peer 192.0.2.2 local-address '192.0.2.1'
set vpn ipsec site-to-site peer 192.0.2.2 vti bind 'vti1'
set vpn ipsec site-to-site peer 192.0.2.2 vti esp-group 'ESP-GRP-VTI'

```
Show sa
```
vyos@r4-1.3# run show vpn ipsec sa

  Invalid command: [IPSec process not running]

[edit]
vyos@r4-1.3# 
```
Logs:
```
May 28 11:53:27 r4-1 ipsec_starter[21795]: Starting strongSwan 5.7.2 IPsec [starter]...
May 28 11:53:27 r4-1 ipsec_starter[21795]: section 'site2' referenced in section 'peer-192.0.2.2-vti' not found
May 28 11:53:27 r4-1 ipsec_starter[21795]: unable to start strongSwan -- fatal errors in config
May 28 11:53:29 r4-1 vyos-configd[19562]: Sending response 1

```
Ipsec conf related vti (**also = site2**)
```
conn peer-192.0.2.2
    authby = secret
    left = 192.0.2.1 # dhcp:no
    right = 192.0.2.2
    keylife = 3600s
    rekeymargin = 540s
    keyexchange = ikev1
    ike = aes256-sha1-modp1024!
    ikelifetime = 3600s
    reauth = no
    closeaction = none

conn peer-192.0.2.2-vti
    also = site2
    leftsubnet = 0.0.0.0/0
    leftupdown = "/etc/ipsec.d/vti-up-down vti1"
    rightsubnet = 0.0.0.0/0
    mark = 9437185
    esp = aes256-sha1-modp1024!
    keylife = 1800s
    compress = no
    type = tunnel

    auto = start
    keyingtries = %forever


```

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Comment by sever-sever** [comment]
> > Impossible to show IKE sa
> > ```
> > vyos@r4-1.3:~$ show vpn ike sa
> > Not implemented
> > vyos@r4-1.3:~$ 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Expected
> > ```
> > vyos@r6-roll:~$ show vpn ike sa
> > Peer ID / IP                            Local ID / IP               
> > ------------                            -------------
> > 100.64.0.1                              100.64.0.2                             
> > 
> >     State  IKEVer  Encrypt  Hash    D-H Group      NAT-T  A-Time  L-Time
> >     -----  ------  -------  ----    ---------      -----  ------  ------
> >     up     IKEv2   aes16_256 n/a     19(ECP_256)    no     3600    86400  
> > 
> >  
> > vyos@r6-roll:~$ 
> > ```
> 
> Oh yes, I was hoping to commit that in a later PR as I've still to understand what is going in the vyatta version: https://github.com/vyos/vyatta-op-vpn/blob/current/lib/OPMode.pm#L1272

Maybe it will be more clear.
https://github.com/vyos/vyatta-op-vpn/pull/27
https://phabricator.vyos.net/T1995

**Comment by sarthurdev** [comment]
Pushed an update.

Fixes:
\- VTI referring to old site naming
\- Copyright to 2021
\- Implemented `show vpn ike sa`
\- Fix `show vpn ipsec sa` when charon not running

**Review by sever-sever — APPROVED** [review]

**Comment by sever-sever** [comment]
> Pushed an update.
> 
> Fixes:
> - VTI referring to old site naming
> - Copyright to 2021
> - Implemented `show vpn ike sa`
> - Fix `show vpn ipsec sa` when charon not running

@sarthurdev  Thank you, you did a great job!

**Comment by sarthurdev** [comment]
> 
> 
> > Pushed an update.
> > Fixes:
> > 
> > * VTI referring to old site naming
> > * Copyright to 2021
> > * Implemented `show vpn ike sa`
> > * Fix `show vpn ipsec sa` when charon not running
> 
> @sarthurdev Thank you, you did a great job!

Thank you!

**Pull request merged**